src/dune_rules/merlin/ocaml_index.ml

```diff
@@ -0,0 +1,109 @@
+open Import
+
+let ocaml_index sctx ~dir =
+  Super_context.resolve_program ~loc:None ~dir sctx "ocaml-index"
+;;
+
+let index_path_in_obj_dir obj_dir =
+  let dir = Obj_dir.obj_dir obj_dir in
+  Path.Build.relative dir "cctx.ocaml-index"
+;;
+
+let project_index ~build_dir = Path.Build.relative build_dir "project.ocaml-index"
+
+let cctx_rules cctx =
+  (* Indexing is performed by the external binary [ocaml-index] which performs
+     full shape reduction to compute the actual definition of all the elements in
+     the typedtree. This step is therefore dependent on all the cmts of those
+     definitions are used by all the cmts of modules in this cctx. *)
+  let sctx = Compilation_context.super_context cctx in
+  let dir = Compilation_context.dir cctx in
+  let aggregate =
+    let obj_dir = Compilation_context.obj_dir cctx in
+    let fn = index_path_in_obj_dir obj_dir in
+    let additional_libs =
+      let open Resolve.Memo.O in
+      let+ non_compile_libs =
+        (* The indexer relies on the load_path of cmt files. When
+           [implicit_transitive_deps] is set to [false] some necessary paths will
+           be missing.These are passed to the indexer with the `-I` flag.
+
+           The implicit transitive libs correspond to the set:
+           (requires_link \ req_compile) *)
+        let* req_link = Compilation_context.requires_link cctx in
+        let+ req_compile = Compilation_context.requires_compile cctx in
+        List.filter req_link ~f:(fun l -> not (List.exists req_compile ~f:(Lib.equal l)))
+      in
+      Lib_flags.L.include_flags non_compile_libs (Lib_mode.Ocaml Byte)
+    in
+    let context_dir =
+      Compilation_context.context cctx
+      |> Context.name
+      |> Context_name.build_dir
+      |> Path.build
+    in
+    let modules_deps =
+      let cm_kind = Lib_mode.Cm_kind.(Ocaml Cmi) in
+      (* We only index occurrences in user-written modules *)
+      Compilation_context.modules cctx
+      |> Modules.With_vlib.drop_vlib
+      |> Modules.fold_user_written ~init:[] ~f:(fun module_ acc ->
+        let cmts =
+          [ Intf; Impl ]
+          |> List.filter_map ~f:(fun ml_kind ->
+            Obj_dir.Module.cmt_file obj_dir ~ml_kind ~cm_kind module_
+            |> Option.map ~f:Path.build)
+        in
+        List.rev_append cmts acc)
+    in
+    Command.run_dyn_prog
+      ~dir:context_dir
```

Comment:
Why do you not run ocaml-index from `dir`? We only run compilation commands from context_dir for good error messages. Does ocaml-index output error messages that refer to source code?

Reply:
`ocaml-index` relies on the load-path stored in the `cmt` files. These paths are relative to the directory in which the compiler was called.

If you would prefer the indexer to run in `dir` we can have dune pass the complete load-path via the CLI.

Reply:
No, that's fine to leave it this way. It makes sense to leave a comment about this though.